cmd/jaeger/internal/sanitizer/utf8.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package sanitizer
+
+import (
+	"unicode/utf8"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/pkg/otelsemconv"
+)
+
+const (
+	invalidOperation = "InvalidOperationName"
+	invalidTagKey    = "InvalidTagKey"
+)
+
+// NewUTF8Sanitizer returns a sanitizer function that performs the following sanitizations
+// on the trace data:
+//   - Checks all attributes of the span to ensure that the keys are valid UTF-8 strings
+//     and all string typed values are valid UTF-8 strings. If any keys or values are invalid,
+//     they are replaced with a valid UTF-8 string containing debugging data related to the invalidations.
+//   - Explicitly checks that all span names are valid UTF-8 strings. If they are not, they are replaced
+//     with a valid UTF-8 string and debugging information is put into the attributes of the span.
+//   - Explicitly checks that the service name is a valid UTF-8 string. If it is not, it is replaced with
+//     a valid UTF-8 string that contains debugging information related to the invalidation.
+func NewUTF8Sanitizer() Func {
+	return sanitizeUTF8
+}
+
+func sanitizeUTF8(traces ptrace.Traces) ptrace.Traces {
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		sanitizeAttributes(resourceSpan.Resource().Attributes())
+
+		scopeSpans := resourceSpan.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			scopeSpan := scopeSpans.At(j)
+			sanitizeAttributes(scopeSpan.Scope().Attributes())
+
+			spans := scopeSpan.Spans()
+
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				if !utf8.ValidString(span.Name()) {
+					sanitized := []byte(span.Name())
+					newVal := span.Attributes().PutEmptyBytes(invalidOperation)
+					newVal.Append(sanitized...)
+					span.SetName(invalidOperation)
+				}
+
+				serviceName, ok := span.Attributes().Get(string(otelsemconv.ServiceNameKey))
```

Comment:
Why do we need to special case service.name? Wouldn't it be handled generically as all other attributes?

Reply:
Also, service name is supposed to be defined at the resource level, not at individual spans.

Reply:
I removed this. It had special handling in v1 but since its just an attribute now it doesn't need any.